include/albatross/src/core/linear_combination.hpp

```diff
@@ -74,6 +74,63 @@ inline auto difference(const X &x, const Y &y) {
   return difference(vx, vy);
 }
 
+template <typename Derived, typename X>
+inline auto operator*(const Eigen::SparseMatrixBase<Derived> &matrix,
+                      const std::vector<X> &features) {
+  std::vector<LinearCombination<X>> output(cast::to_size(matrix.rows()));
+
+  std::vector<std::vector<double>> output_coefs(cast::to_size(matrix.rows()));
+
+  for (Eigen::Index k = 0; k < matrix.outerSize(); ++k) {
+    for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
+      output[cast::to_size(it.row())].values.push_back(
+          features[cast::to_size(it.col())]);
+      output_coefs[cast::to_size(it.row())].push_back(it.value());
+    }
+  }
+
+  for (std::size_t i = 0; i < output_coefs.size(); ++i) {
```

Comment:
Can we just output directly into the coefficients of the `LinearCombination`s above instead of doing another copy here?  Might need to reserve the right size for the coefficient vectors ahead of time?

Reply:
I gave that a shot, it'll avoid copying the `.data()` from a `vector<double>` into the matrix ... but I'm pretty sure it introduces another loop including static casting and I'm unsure whether that means another loop of copying (if so I think the new version will be less efficient right?).

Reply:
Actually, I'm going to scrap my modified version. it doesn't pass tests what I tried was something like:
```
  for (Eigen::Index k = 0; k < matrix.outerSize(); ++k) {
    Eigen::Index elem_count = 0;
    for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
      const auto srow = cast::to_size(it.row());
      const auto scol = cast::to_size(it.col());
      output[srow].values.push_back(features[scol]);
      ++elem_count;
    }

    output[cast::to_size(it.row())].coefficients.resize(elem_count);
    Eigen::Index j = 0;
    for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
      output[it.row()].coefficients[j] = it.value();
      ++j
    }
  }
```
but notice that outer `k` loop isn't a loop over rows necessarily, we need to access `it.row()` and `it.col()` which means I'd have to either 1) make some assumptions about the way the sparse matrix is stored which aren't clearly held here (see ["iterating over non-zero elements"](https://eigen.tuxfamily.org/dox/group__TutorialSparse.html)) or 2) keep a bookkeeping vector around in which I increment the element counts in each row.